### PR TITLE
[expo-dev-launcher][android] add ability to load published projects via expo-updates

### PR DESCRIPTION
**Pull request by esamelson**
# Why

step 2 of ENG-957 and ENG-958

This PR adds a dependency on `expo-updates-interface` to `expo-dev-launcher` and all the required logic to use the provided interface to load a published project via the provided interface into expo-updates.

The actual implementation of the `UpdatesInterface` must be provided externally to dev-launcher (i.e. set up in MainApplication), and we handle the case where no implementation is provided.

# How

- Since most of the logic from both `DevLauncherAppLoader` and `DevLauncherExpoAppLoader` could be reused between published and locally served apps, I made `DevLauncherExpoAppLoader` an abstract class with both `Published` and `Local` implementations.  Rather than an abstract `getBundleUrl` method, `DevLauncherAppLoader` now has an abstract `injectBundleLoader` method since its implementation differs between locally served + published projects.
- In MainApplication's ReactNativeHost, the return value of `getUseDeveloperSupport` now needs to depend on whether the bundle we're loading is published or locally served -- otherwise we can get the built-in RN dev menu in published projects by shaking the device. I've added a field to DevLauncherController to support that, but it does require another small diff in the installation/setup instructions.
- I removed the call to `changeUrlScheme(url, "http")` since we no longer know before receiving a URL whether its scheme should be `http` or `https`.
- The `HEAD` request in the `isManifestUrl` check is a bit problematic, since the classic updates server responds with a 400 if the correct headers aren't provided in the request. I solved this for now by returning `true` if the response is unsuccessful, but I'm not sure this is the best solution long-term. (Do we want to keep this functionality that allows loading directly from `react-native start`? As far as I can tell it's broken right now anyway.)
- Added `@` as an allowed character in the URL validator so we can load published projects 🙂 also turned off auto-capitalize as it's unnecessary and annoying for a URL input field.

# Test Plan

Manual tests in #13032.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

# Todo

We'll need to update the installation instructions to include setting the `updatesInterface` property, and reading from the `useDeveloperSupport` property.

### CODE REVIEW

File: packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/helpers/DevLauncherReactUtils.kt [hunk 1]
```diff
@@ -56,13 +69,41 @@ fun injectDebugServerHost(
...
+    mUseDeveloperSupportField.isAccessible = true
+    mUseDeveloperSupportField[instanceManager] = false
+    true
+  } catch (e: Exception) {
+    Log.e("DevLauncher", "Unable to load local bundle file", e)
+    false
+  }
+}
+
+
```
Comment (lukmccall) [inline packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/helpers/DevLauncherReactUtils.kt:106]:
```suggestion
```

File: packages/expo-dev-launcher/package.json [hunk 2]
```diff
@@ -28,7 +28,8 @@
   "license": "MIT",
   "homepage": "https://docs.expo.io",
   "dependencies": {
-    "@expo/config-plugins": "^1.0.32"
+    "@expo/config-plugins": "^1.0.32",
+    "expo-updates-interface": "~1.0.0"
```
Comment (lukmccall) [inline packages/expo-dev-launcher/package.json:32]: We don't want to have this interface here.

**Review by lukmccall — CHANGES_REQUESTED** [review]
Generally looks very solid 🚀
I've left some comments 😉

File: packages/expo-dev-launcher/package.json (hunk 2 above)
Comment (esamelson) [reply packages/expo-dev-launcher/package.json:32]: Do we want it in the expo-dev-client package instead? Either way, I'm curious why we don't want this here; the code will not compile without this dependency.

Comment (lukmccall) [reply packages/expo-dev-launcher/package.json:32]: Yes, we want to add it to the expo-dev-client package instead. But also to the `dependencies.js` file to link it automatically to the react-native project. See https://github.com/expo/expo/blob/6d31ade729586695bf9aaf68c04367aa4d68190b/packages/expo-dev-client/dependencies.js#L3-L13

We don't want to have it here cause even with it, it won't compile. The react-native won't link it.

**Review by lukmccall — APPROVED** [review]
Resolve https://github.com/expo/expo/pull/13031/files#r640501621 and I think we can merge it ;)

**Pull request merged**